packages/core/upload/server/src/utils/mime-validation.ts

```
@@ -0,0 +1,279 @@
+import { fileTypeFromBuffer } from 'file-type';
+import { readFile } from 'node:fs/promises';
+import type { Core } from '@strapi/types';
+import { errors } from '@strapi/utils';
+
+export type SecurityConfig = {
+  allowedTypes?: string[];
+  deniedTypes?: string[];
+};
+type UploadValidationError = {
+  code: 'MIME_TYPE_NOT_ALLOWED' | 'VALIDATION_ERROR' | 'UNKNOWN_ERROR';
+  message: string;
+  details: Record<string, any>;
+};
+
+type ValidationResult = {
+  isValid: boolean;
+  error?: UploadValidationError;
+};
+
+type ErrorDetail = {
+  file: any;
+  originalIndex: number;
+  error: UploadValidationError;
+};
+
+async function readFileChunk(filePath: string, chunkSize: number = 4100): Promise<Buffer> {
+  const buffer = await readFile(filePath);
+  return buffer.length > chunkSize ? buffer.subarray(0, chunkSize) : buffer;
+}
+
+export async function detectMimeType(file: any): Promise<string | undefined> {
+  try {
+    let buffer: Buffer;
+
+    const filePath = file.path || file.filepath || file.tempFilePath;
+
+    if (filePath) {
+      buffer = await readFileChunk(filePath, 4100);
+    } else if (file.buffer) {
+      buffer = file.buffer.length > 4100 ? file.buffer.subarray(0, 4100) : file.buffer;
+    } else {
+      return undefined;
+    }
+
+    const result = await fileTypeFromBuffer(new Uint8Array(buffer));
+
+    return result?.mime;
+  } catch (error) {
+    return undefined;
+  }
```

How come we fail silently?